js_modules/dagit/packages/core/src/assets/AssetMaterializations.tsx

```diff
@@ -41,66 +40,121 @@ interface Props {
   // This timestamp is a "hint", when it changes this component will refetch
   // to retrieve new data. Just don't want to poll the entire table query.
   assetLastMaterializedAt: string | undefined;
+  assetHasDefinedPartitions: boolean;
 }
 
 const LABEL_STEP_EXECUTION_TIME = 'Step Execution Time';
 
+/**
+ * If the asset has a defined partition space, we load all materializations in the
+ * last 200 partitions. This ensures that if you run a huge backfill of old partitions,
+ * you still see accurate info for the last 200 partitions in the UI. A count-based
+ * limit could cause random partitions to disappear if materializations were out of order.
+ *
+ * For non-SDA-partitioned assets, we load the most recent 200 materializations. We might
+ * still show these "By partition" (and the gaps problem above exists), but we don't have
+ * a choice.
+ */
+function useRecentMaterializations(
+  assetKey: AssetKey,
+  assetHasDefinedPartitions: boolean,
+  xAxis: 'partition' | 'time',
+  before?: string,
+) {
+  const loadUsingPartitionKeys = assetHasDefinedPartitions && xAxis === 'partition';
+
+  const {data, loading, refetch} = useQuery<
+    AssetMaterializationsQuery,
+    AssetMaterializationsQueryVariables
+  >(ASSET_MATERIALIZATIONS_QUERY, {
+    variables: loadUsingPartitionKeys
+      ? {
+          assetKey: {path: assetKey.path},
+          before: before,
+          partitionInLast: 120,
+        }
+      : {
+          assetKey: {path: assetKey.path},
+          before: before,
+          limit: 200,
+        },
+  });
+
+  const asset = data?.assetOrError.__typename === 'Asset' ? data?.assetOrError : null;
+  const materializations = React.useMemo(() => asset?.assetMaterializations || [], [asset]);
+  const allPartitionKeys = asset?.definition?.partitionKeys;
+  const requestedPartitionKeys =
+    loadUsingPartitionKeys && allPartitionKeys
+      ? allPartitionKeys.slice(allPartitionKeys.length - 120)
+      : undefined;
+
+  return {asset, requestedPartitionKeys, materializations, loading, refetch};
```

Comment:
`return React.useMemo(() => ({asset,  ...}, [...])` to avoid having to create this object anew every time? 🤷